### PR TITLE
plugin/rewrite: rewrite responses for all record types (#1460)

### PR DESCRIPTION
**Pull request by greenpau**
### 1. What does this pull request do?

It add the ability to rewrite answers for all RR types, not only A records.

### 2. Which issues (if any) are related?

#1460 

### 3. Which documentation changes (if any) need to be made?

None.

### CODE REVIEW

**Comment by greenpau** [comment]
@miekg , please review.

**Comment by miekg** [comment]
lgtm, but as this fixes a bug; it should have a test case checking for this.

**Comment by greenpau** [comment]
> lgtm, but as this fixes a bug; it should have a test case checking for this.

@miekg , fair point 😃

**Comment by greenpau** [comment]
@miekg , regarding the tests. I am planning to create `reverter_test.go`. 

I came to like [Ginkgo](https://github.com/onsi/ginkgo). Unfortunately, I did not develop enough proficiency in using it, yet. Do you mind if I use it for the testing here?

**Comment by miekg** [comment]
`reverter_test.go` sgtm.

Please refrain from using external testing frameworks, the normal go stuff should be good enough.

**Comment by greenpau** [comment]
> Please refrain from using external testing frameworks, the normal go stuff should be good enough.

👍

**Comment by greenpau** [comment]
@miekg , please take a look at the screenshot below. Considering my tests do response rewrites, why does Codecov show no tests against that section of code?

screen shot 2018-02-06 at 12 32 35 am

**Comment by greenpau** [comment]
@miekg , when I run `go test -v plugin/rewrite/*.go` and add extra debugging, I can clearly see that my test code hits all the way to line 44. However, Codecov is not showing it. Any hints?

screen shot 2018-02-06 at 1 37 37 am

**Comment by miekg** [comment]
Ack. I'll double check by checking out the PR locally. But I wouldn't worry too much about codecov; it gets more things wrong.

File: plugin/rewrite/reverter_test.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	"github.com/coredns/coredns/plugin/pkg/dnstest"
+	"github.com/coredns/coredns/plugin/test"
+
+	"github.com/miekg/dns"
+	"golang.org/x/net/context"
+)
+
+func TestResponseReverter(t *testing.T) {
+	rules := []Rule{}
+	r, _ := newNameRule("stop", "regex", "(core)\\.(dns)\\.(rocks)", "{2}.{1}.{3}", "answer", "name", "(dns)\\.(core)\\.(rocks)", "{2}.{1}.{3}")
```
Comment (miekg) [inline plugin/rewrite/reverter_test.go]: use backticks for the regexps: "(core)\\.(dns)\\.(rocks)" -> `(core)\.(dns)\.(rocks)` (markdown doesn't show - but the later uses `)

File: plugin/rewrite/reverter_test.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+		from     string
+		fromType uint16
+		answer   []dns.RR
+		to       string
+		toType   uint16
+		noRevert bool
+	}{
+		{"core.dns.rocks", dns.TypeA, []dns.RR{test.A("dns.core.rocks.  5   IN  A   10.0.0.1")}, "core.dns.rocks", dns.TypeA, false},
+		{"core.dns.rocks", dns.TypeSRV, []dns.RR{test.SRV("dns.core.rocks.  5   IN  SRV 0 100 100 srv1.dns.core.rocks.")}, "core.dns.rocks", dns.TypeSRV, false},
+		{"core.dns.rocks", dns.TypeA, []dns.RR{test.A("core.dns.rocks.  5   IN  A   10.0.0.1")}, "dns.core.rocks.", dns.TypeA, true},
```
Comment (miekg) [inline plugin/rewrite/reverter_test.go]: Maybe add a whacky older RR so we test that anything hitting this gets rewritten correctly? I.e a `HINFO` record?

**Review by miekg — APPROVED** [review]
some nits, lgtm.

File: plugin/rewrite/reverter_test.go (hunk 2 above)
Comment (greenpau) [reply plugin/rewrite/reverter_test.go]: >  I.e a HINFO record?

@miekg , had to do some 🏋️ in `plugin/test` 😆

**Comment by greenpau** [comment]
> some nits, lgtm.

@miekg , please 👓

File: plugin/rewrite/reverter_test.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+func TestResponseReverter(t *testing.T) {
+	rules := []Rule{}
+	r, _ := newNameRule("stop", "regex", `(core)\.(dns)\.(rocks)`, "{2}.{1}.{3}", "answer", "name", `(dns)\.(core)\.(rocks)`, "{2}.{1}.{3}")
+	rules = append(rules, r)
+
+	tests := []struct {
+		from     string
+		fromType uint16
+		answer   []dns.RR
```
Comment (miekg) [inline plugin/rewrite/reverter_test.go:22]: this doesn't need to be a slice, we only ever put 1 RR in it, did you intent to test multiple?

File: plugin/test/helpers.go [hunk 4]
```diff
@@ -81,6 +81,9 @@ func PTR(rr string) *dns.PTR { r, _ := dns.NewRR(rr); return r.(*dns.PTR) }
 // TXT returns a TXT record from rr. It panics on errors.
 func TXT(rr string) *dns.TXT { r, _ := dns.NewRR(rr); return r.(*dns.TXT) }
 
+// HINFO returns a HINFO record from rr. It panics on errors.
+func HINFO(rr string) *dns.HINFO { r, _ := dns.NewRR(rr); return r.(*dns.HINFO) }
```
Comment (miekg) [inline plugin/test/helpers.go:85]: ha, we didn't have it. Nice. OK :)

File: plugin/rewrite/reverter_test.go (hunk 3 above)
Comment (greenpau) [reply plugin/rewrite/reverter_test.go:22]: > did you intent to test multiple?

@miekg , yes, I did. Good point. I will add that test case.

As a side note, I was wondering about multiple names in a single request. I don't think the plugin supports it at the moment.

**Comment by greenpau** [comment]
@miekg, please review.

File: plugin/rewrite/reverter_test.go (hunk 3 above)
Comment (miekg) [reply plugin/rewrite/reverter_test.go:22]: there can only be one name in the question section in DNS

Comment (greenpau) [reply plugin/rewrite/reverter_test.go:22]: >  there can only be one name in the question section in DNS

@miekg , I see 👍 . I was self-educating myself the other day and stumbled upon [stack thread](
https://stackoverflow.com/questions/4082081/requesting-a-and-aaaa-records-in-single-dns-query/4083071#4083071). Then, kept it in memory to ask someone about it 😃

**Pull request merged**